### PR TITLE
8198317: Enhance JavacTool.getTask for flexibility

### PR DESCRIPTION
**Pull request by lahodaj**
This is a clone of https://github.com/openjdk/jdk/pull/1896, which got closed by Skara.

@lgxbslgx, please let me know if you would prefer to resubmit the PR yourself.

/contributor add lgxbslgx

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8198317](https://bugs.openjdk.java.net/browse/JDK-8198317): Enhance JavacTool.getTask for flexibility

### Reviewers
 * [Joel Borggrén-Franck](https://openjdk.java.net/census#jfranck) (@jbf - **Reviewer**) ⚠️ Review applies to ec3c8927179f86d1da76e4648512820d04aca9bb

### Contributors
 * Guoxiong Li `<lgxbslgx@gmail.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2871/head:pull/2871` \
`$ git checkout pull/2871`

Update a local copy of the PR: \
`$ git checkout pull/2871` \
`$ git pull https://git.openjdk.java.net/jdk pull/2871/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2871`

View PR using the GUI difftool: \
`$ git pr show -t 2871`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2871.diff">https://git.openjdk.java.net/jdk/pull/2871.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/api/8198317/T8198317.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import javax.tools.JavaFileObject;
+
+import com.sun.tools.javac.api.JavacTaskImpl;
+import com.sun.tools.javac.util.Log;
+
+import toolbox.ToolBox;
+import toolbox.TestRunner;
+import static toolbox.ToolBox.lineSeparator;
+
+public class T8198317 extends TestRunner{
```
Comment (liach) [inline test/langtools/tools/javac/api/8198317/T8198317.java:54]: formatting

Comment (jonathan-gibbons) [inline test/langtools/tools/javac/api/8198317/T8198317.java:54]: These days, it is considered bad form to base a test name on a bug number. It's better to come up with a meaningful name for the test.  That probably applies to the enclosing directory as well.

**Review by jonathan-gibbons — CHANGES_REQUESTED** [review]

**Comment by lgxbslgx** [comment]
@lahodaj It is good for me to continue the work in this PR. I will assist you if you need.

File: test/langtools/tools/javac/api/TestContextLoggingOutput.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (lgxbslgx) [inline test/langtools/tools/javac/api/TestContextLoggingOutput.java:2]: The copyright should be updated to 2021.

**Review by lgxbslgx — COMMENTED** [review]
The [original patch](https://github.com/openjdk/jdk/pull/1896) began at Dec 2020. About four months passed. Could somebody help me push this patch? Thanks a lot.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTool.java [hunk 3]
```diff
@@ -166,9 +166,15 @@ public JavacTask getTask(Writer out,
             if (diagnosticListener != null)
                 context.put(DiagnosticListener.class, ccw.wrap(diagnosticListener));
 
-            if (out == null)
+            // If out is null and the value is set in the context, we need to do nothing.
+            if (out == null && context.get(Log.errKey) == null)
+                // Situation: out is null and the value is not set in the context.
                 context.put(Log.errKey, new PrintWriter(System.err, true));
-            else
+            else if (out instanceof PrintWriter)
```
Comment (jbf) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTool.java:173]: Can use instance of pattern matching

**Review by jbf — APPROVED** [review]
Looks good to me

**Pull request closed without merging**